### ISSUE TITLE
Incorrect type narrowing in negative case for `isinstance` when filter type is generic

### ISSUE THREAD

**Issue by robert-bn** [microsoft/pyright#7081]
**Description**
When using a union of generic type parameters, and `isinstance` to narrow bounds, impossible cases are not identified as impossible.

**Code**
In the function below
```python
from typing import assert_never

def test1[T,U](
    cls_t: type[T],
    cls_u: type[U],
    x: T | U,
):
    if isinstance(x, cls_t) or isinstance(x, cls_u):
        return
    assert_never(x)
```
pyright reports a type error due to assert_never potentially being reachable:
```
error: Argument of type "object*" cannot be assigned to parameter "__arg" of type "Never" in function "assert_never"
    Type "object*" cannot be assigned to type "Never" (reportArgumentType)
```
As x is type `T | U`, at least one of `isinstance(x, cls_t)` and `isinstance(x, cls_u)` must be true, hence `assert_never(x)` can never be reached.

Interestingly,
```python
def test2[T,U](
    cls_t: type[T],
    cls_u: type[U],
    x: T | U,
):
    if isinstance(x, (cls_t, cls_u)):
        return
    assert_never(x)
```
which is equivalent, results in no type error.

Checked with pyright 1.1.348 command line, and vscode pylance v2023.12.1.

My apologies if this is as designed or has already been reported.

**Comment by erictraut** [comment]
I agree there's a bug here, but it's not what you're reporting. I think pyright is being too aggressive in type narrowing here. 

Consider the following simplified example:
```python
def test(cls_t: type[T], x: T) -> T:
    if isinstance(x, cls_t):
        reveal_type(x)  # Pyright reveals "object & T", mypy reveals "T"
        return x
    reveal_type(x)  # Pyright reveals "Never", mypy reveals "T"
    print("Not Never 0!")
    return x

result = test(int, "")
reveal_type(result)  # Pyright reveals "int | str", mypy reveals "object"
```

So pyright has a bug where it is incorrectly narrowing to `Never` in the negative ("else") case. The type should remain `T` in this path. Mypy gets the right answer.

**Comment by robert-bn** [comment]
> I agree there's a bug here, but it's not what you're reporting. I think pyright is being too aggressive in type narrowing here.

Ah yes, I incorrectly identified the impossibility of reaching `assert_never` since  we can always call with a subtype of `T` that doesn't overlap with the type of x.  Thank you for helping my understanding there.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.350, which I just published.
